### PR TITLE
Remove State.GetModel()

### PR DESCRIPTION
**Pull request by mjs**
## Description of change

These changes remove the `GetModel` method from `state.State`. This gets rid of the last remaining problematic way for `state.Model` instances to be created. `Model` instances can now only be created using `State.Model()` which means the `State` a `Model` references is now guaranteed to be connected to the same model UUID as the `Model`. This paves the way for moving more functionality from `State` onto `Model` and is a prerequisite for upcoming feature work.

A follow-up PR will clean up the internals of `state.Model`, taking advantage of the fact that we know a `Model`'s `State` is now the correct one.

This change has had many positive follow-on effects. More uses of the evil `State.ForModel()` were removed and the code and tests for some apiserver facades is now significantly cleaner. Some minor inefficiencies have also been dealt with.

## QA steps

The areas at risk of breakage as a result of this change are model/controller creation and destruction, and model migrations, so exercise those:

```
$ juju bootstrap lxd source
# wait...
$ juju bootstrap lxd target
# wait...
$ juju switch source
$ juju add-model foo
$ juju deploy ubuntu
# wait...

# Check for significant errors
$ juju debug-log -m source:foo--replay --no-tail -l ERROR
$ juju debug-log -m source:controller --replay --no-tail -l ERROR

# Migrate
$ juju migrate foo target

# Wait for model to appear on target
$ watch juju status -m target:foo

# Add another unit
$ juju add-unit ubuntu
# wait...

# Check for significant errors
$ juju debug-log -m source:controller --replay --no-tail -l ERROR
$ juju debug-log -m target:controller --replay --no-tail -l ERROR
$ juju debug-log -m target:foo --replay --no-tail -l ERROR

# Destroy controllers
$ juju destroy-controller -y --destroy-all-models source
$ juju destroy-controller -y --destroy-all-models target
```

## Documentation changes

N.A. - internal only

## Bug reference

N.A.

### CODE REVIEW

File: apiserver/apiserver.go [hunk 1]
```diff
@@ -1060,12 +1060,13 @@ func (srv *Server) processModelRemovals() error {
...
 		case modelUUIDs := <-w.Changes():
 			for _, modelUUID := range modelUUIDs {
-				model, err := st.GetModel(names.NewModelTag(modelUUID))
+				model, release, err := srv.statePool.GetModel(modelUUID)
 				gone := errors.IsNotFound(err)
 				dead := err == nil && model.Life() == state.Dead
 				if err != nil && !gone {
 					return errors.Trace(err)
 				}
+				release()
```
Comment (wallyworld) [inline apiserver/apiserver.go]: At this point, err could be not found. Will release() be valid then? Don't we want to check that err == nil?

Comment (mjs) [inline apiserver/apiserver.go]: Good point. Will fix.

File: apiserver/facades/client/machinemanager/state.go [hunk 2]
```diff
@@ -44,47 +59,20 @@ func (s stateShim) Machine(name string) (Machine, error) {
...
-type Machine interface {
-	Destroy() error
-	ForceDestroy() error
-	Units() ([]Unit, error)
+func (p *poolShim) GetModel(uuid string) (Model, func(), error) {
+	m, release, err := p.pool.GetModel(uuid)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return m, func() { release() }, nil
```
Comment (babbageclunk) [inline apiserver/facades/client/machinemanager/state.go]: can't this just be `return m, release, nil`?

File: apiserver/facades/controller/modelupgrader/backend.go [hunk 3]
```diff
@@ -21,18 +23,18 @@ type Model interface {
...
-func (s stateBackend) GetModel(tag names.ModelTag) (Model, error) {
-	m, err := s.State.GetModel(tag)
+func (p *statePool) GetModel(uuid string) (Model, func(), error) {
+	m, release, err := p.pool.GetModel(uuid)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, nil, errors.Trace(err)
 	}
-	return m, nil
+	return m, func() { release() }, nil
```
Comment (babbageclunk) [inline apiserver/facades/controller/modelupgrader/backend.go]: Same here.

File: apiserver/facades/client/machinemanager/state.go (hunk 2 above)
Comment (mjs) [inline apiserver/facades/client/machinemanager/state.go]: Nope. The release func returned by GetModel is `func () bool` but the Pool interface simplifies that to `func()`

File: apiserver/facades/client/controller/controller_test.go [hunk 4]
```diff
@@ -123,11 +123,12 @@ func (s *controllerSuite) TestAllModels(c *gc.C) {
...
 	var obtained []string
-	for _, env := range response.UserModels {
-		obtained = append(obtained, env.Name)
-		stateEnv, err := s.State.GetModel(names.NewModelTag(env.UUID))
+	for _, userModel := range response.UserModels {
+		obtained = append(obtained, userModel.Name)
+		stateModel, release, err := s.StatePool.GetModel(userModel.UUID)
 		c.Assert(err, jc.ErrorIsNil)
-		s.checkEnvironmentMatches(c, env.Model, stateEnv)
+		defer release()
```
Comment (mjs) [inline apiserver/facades/client/controller/controller_test.go]: I guess `checkEnvironmentMatches` doesn't call anything on the model which relies on the DB so calling `release` here is safe enough... probably not worth it though.

File: apiserver/facades/controller/modelupgrader/backend.go (hunk 3 above)
Comment (mjs) [inline apiserver/facades/controller/modelupgrader/backend.go]: Same as above

**Comment by mjs** [comment]
$$merge$$

**Pull request merged**